Add platform developer access policy (#630)
opensafely/documentation: docs/developer-access-policy.md

# Policy for OpenSAFELY Access by Platform Developers

## Introduction

OpenSAFELY securely mediates access to pseudonymized patient data for the purpose of analysis (be that research, service evaluation or audit).
Our objective is to minimise direct access to pseudonymised data wherever possible, to protect privacy;
while facilitating high quality, efficient analytics on the data by a wide range of users.

> A small number of OpenSAFELY team members need more direct access to the raw pseudonymised data inside the secure environment

Reviewer: Nit: We could remove "raw" here, IMHO. We don't use it in the first sentence (above) and, thinking defensively, "raw" might seem to contradict "pseudonymized".

> A small number of OpenSAFELY team members need more direct access to the raw pseudonymised data inside the secure environment
> for the purposes of data curation and for developing and running the OpenSAFELY platform.

Reviewer 2: ```suggestion
A small number of OpenSAFELY team members need more direct access to the raw pseudonymised data inside the secure environment for:

* developing and running the OpenSAFELY platform
* data curation purposes 
```

I've also switched these two points around to match the document order.

> This policy describes the conditions under which individuals with these roles may access the OpenSAFELY platform
> and the controls that are in place to ensure that such access cannot be used to subvert the OpenSAFELY security model.
> All such access conforms to the OpenSAFELY platform’s principles of security, transparency and minimal access.
> 
> The specific types of access activities this policy describes are:
> 
> * Running OpenSAFELY test pipelines
> * Ad hoc access to backend systems
> * Data curation activities

Reviewer 2: ```suggestion
### What this policy describes

This policy describes:

* the conditions under which individuals with these roles may access the OpenSAFELY platform
* the controls in place to ensure that such access cannot be used to subvert the OpenSAFELY security model

All such access conforms to the OpenSAFELY platform’s principles of security, transparency and minimal access.

The specific types of access activities this policy covers are:

* Running OpenSAFELY test pipelines
* Ad hoc access to backend systems
* Data curation activities
```

## Running OpenSAFELY test pipelines

### Why

> For software maintenance and service provision, it is sometimes necessary to execute OpenSAFELY [study pipelines](https://docs.opensafely.org/actions-pipelines/#project-pipelines).

Reviewer 2: suggestion:

The link pointed to has a heading of "project pipelines". But the link text is "study pipelines". I'd personally go for consistency and use "project pipelines" here. [^1]

[^1]: If the correct term really is "study pipelines", we should change that everywhere later.

Reviewer 2: comment:

This sentence sounds like it is now talking about both normal running — "service provision" — of OpenSAFELY and testing. Perhaps it's simpler to solely focus on the "software maintenance" aspect? That is what is being discussed. This means you can rewrite as:

> For software maintenance purposes, it is sometimes necessary to run OpenSAFELY project pipelines.

> In particular this is necessary to ensure that the platform as a whole is operating correctly and to verify that changes to the software work correctly in the deployed system.

Reviewer 2: ```suggestion
The test project pipelines are necessary to:

* ensure that the entire OpenSAFELY platform is operating correctly
* verify that changes to the software work correctly in deployed systems
```

### What

The OpenSAFELY team maintains a small set of test studies in the [OpenSAFELY Testing](https://jobs.opensafely.org/datalab/opensafely-testing/) project.
These studies have pipelines that use the standard OpenSAFELY tooling,
but which have been specially designed to test and assess the platform’s functionality rather than to answer research questions.

Because they are standard OpenSAFELY pipelines, they benefit from all the same security controls as OpenSAFELY research pipelines,

> including publication of all code that is run, logging of all operations and manual checking of outputs for disclosivity before they are released from the backend.

Reviewer 2: ```suggestion
including:

* publication of all code that is run
* logging of all operations
* manual checking of outputs for disclosivity before they are released from the backend

```

Access to the OpenSAFELY platform will also be required to verify the pipeline outputs are as expected.

### Who

The test studies can only be run by OpenSAFELY Core Developers.

### When

These test pipelines are run periodically as is necessary to check that the system is working properly.
They are also run when updated software has been deployed;

> however this deployment only happens after new versions of the software have gone through careful testing with simulated data to ensure that they are functioning correctly and that the security characteristics of the platform are maintained.

Reviewer 2: question: Is this a necessary detail? It might be, it might not. The previous two sentences read fine to me without it.

I'd take it as given that we aim to maintain the security promises of the platform. But maybe it does need to be stated?

Reviewer 2: If it's necessary, I'd replace the semi-colon on previous line with a full stop.

Author: I think the detail is helpful. Since this is (at least partly) for testing purposes, it's good to reassure people that we're not just throwing any old code up there, but code that has already been validated. I'll split the sentence.

> ## _Ad hoc_ access to backend systems

Reviewer: Nit: Inconsistent italicization. IMHO we don't need to italicize it, because it's a word that's now entered the English language and is no longer on loan from Latin.

> ## _Ad hoc_ access to backend systems

Reviewer 2: ```suggestion
## Ad hoc access to backend systems
```

I'd remove the italics for consistency with the usage above. From reading around, it tends not to be italicised.

> Note: this section applies to OpenSAFELY-TPP and OpenSafely-EMIS only.

Reviewer 2: ```suggestion
Note: this section applies to OpenSAFELY-TPP and OpenSAFELY-EMIS only.
```

### Why

> In order to ensure the correct operation of the platform,
> it is sometimes necessary to directly access OpenSAFELY backend systems so that software can be reconfigured or updated and problems which only manifest themselves in the deployed system can be investigated.

Reviewer 2: ```suggestion
To ensure the correct operation of the platform,
it is sometimes necessary to directly access OpenSAFELY backend systems so that:

* software can be reconfigured or updated
* problems which only manifest themselves in the deployed system can be investigated
```

### What

Access to backend systems is via secure, authenticated, encrypted channels. All access to the system and database operations are logged.
This process gives access
to the OpenSAFELY software running in the backend and logs that that software produces,
to pseudonymized patient data,
to unvetted intermediate data

> and to unchecked, unreleased study outputs.

Reviewer: Nit: This is a case of the text reading better in source than in rendered versions. This is an inline list, where each list item begins "to...". Possibly separate list items with semicolons rather than commas?

```
This process gives access
to the OpenSAFELY software running in the backend and to logs that that software produces;
to pseudonymized patient data;
to unvetted intermediate data;
and to unchecked, unreleased study outputs.
```

> This process gives access
> to the OpenSAFELY software running in the backend and logs that that software produces,
> to pseudonymized patient data,
> to unvetted intermediate data
> and to unchecked, unreleased study outputs.

Reviewer 2: ```suggestion
This process gives access to:

* the OpenSAFELY software running in the backend
* logs that the OpenSAFELY software produces
* pseudonymized patient data
* unvetted intermediate data
* unchecked, unreleased study outputs.
```

> There is no facility for extracting data from the system via this process, it can only be studied within the system itself.

Reviewer 2: ```suggestion
There is no facility for extracting data from the system via this process: the data can only be studied within the system itself.
```

### Who

Direct access to the OpenSAFELY backend systems is only available to OpenSAFELY Core Developers.

### When

This process is routinely used for updating the platform with new software versions.
It is exceptionally used for investigating outages or other problems, and only when the information necessary for the investigation can’t be accessed from outside the system.

## Data curation activities

> Note: this section applies to OpenSAFELY-TPP and OpenSafely-EMIS only.

Reviewer 2: ```suggestion
!!! note
    This section applies to OpenSAFELY-TPP and OpenSafely-EMIS only.
```

### Why

> Characterizing new and updated datasets, or individual fields which were present but not previously in use, checking data quality and sense-checking.

Reviewer 2: nit: This is slightly tricky to read for me: is "sense-checking" part of "checking data quality"? "sense-checking" might also be a little ambiguous.

Author: I've radically simplified this since most of it is duplicated in the "when" section below.

### What

Access to backend systems is via secure, authenticated, encrypted channels. All access to the system and database operations are logged.
Because data curation requires patterns of data access that are not supported by the standard OpenSAFELY pipelines,
this process allows custom code to be run with access to the pseudonymized patient data, outside the pipelines.

These data curation activities benefit from all the same security controls as standard OpenSAFELY pipelines,

> including publication of all code that is run, logging of all operations and manual checking of outputs for disclosivity before they are released from the backend.

Reviewer 2: ```suggestion
including:

* publication of all code that is run
* logging of all operations
* manual checking of outputs for disclosivity before they are released from the backend
```

### Who

> Only OpenSAFELY Core Developers and OpenSAFELY Data Investigators have permission to carry out data curation activities.

Reviewer 2: question: Do we need to define these roles somewhere — possibly the glossary? — if they are not defined already?

I don't remember ever seeing "Data Investigator". From a search, "core" is used in a couple of ambiguous ways in the documentation; for example, "core OpenSAFELY team".

Author: They are roles defined in the job server. There is an [outstanding issue](https://github.com/opensafely-core/job-server/issues/1624) to document them completely. And once this PR is merged I'll add a note to that issue to put hyperlinks from this policy to the definitions.

### When

> Data curation activities may be carried out when new datasets are added to OpenSAFELY or datasets are updated, and periodically for data quality assurance and sense checking purposes.

Reviewer 2: nit: This might be slightly easier to follow as a bulleted list:

```suggestion
Data curation activities may be carried out:

* when new datasets are added to OpenSAFELY
* when datasets are updated
* periodically for data quality assurance and sense checking purposes
```
